File: Python/oomph-data-ws/uf_309_test/ufs/v321/ABC328_F.py
# https://atcoder.jp/contests/abc328/tasks/abc328_f

from ....uf_309_lib.ufs.int_omega.unionfind_321 import UnionFind

n, qc = map(int, input().split())

r = []
uf = UnionFind(n + 1, 0)

for i in range(qc):
    a, b, d = map(int, input().split())
    if uf.union(b, a, d) or uf.verify(b, a, d):
        r.append(i + 1)

print(*r)

File: CSharp/Oomph.Data/UF09Test/UFs/v321/ABC328_F.cs
﻿using System;
using System.Collections.Generic;
using System.Linq;
using Oomph.Data.UF09Lib.UFs.v321;

namespace UF09Test.UFs.v321
{
	// Test: https://atcoder.jp/contests/abc328/tasks/abc328_f
	class ABC328_F
	{
		static int[] Read() => Array.ConvertAll(Console.ReadLine().Split(), int.Parse);
		static (int, int) Read2() { var a = Read(); return (a[0], a[1]); }
		static (int, int, int) Read3() { var a = Read(); return (a[0], a[1], a[2]); }
		static void Main() => Console.WriteLine(Solve());
		static object Solve()
		{
			var (n, qc) = Read2();
			var qs = Array.ConvertAll(new bool[qc], _ => Read3());

			var r = new List<int>();
			var uf = new UnionFind<long>(n + 1, 0, x => -x, (x, y) => x + y);

			for (int i = 0; i < qc; i++)
			{
				var (a, b, d) = qs[i];

				if (uf.Union(b, a, d) || uf.Verify(b, a, d))
				{
					r.Add(i + 1);
				}
			}
			return string.Join(" ", r);
		}
	}
}
